odoo/models.py

```diff
@@ -5590,35 +5590,66 @@ def exists(self):
         valid_ids = set([r[0] for r in self._cr.fetchall()] + new_ids)
         return self.browse(i for i in self._ids if i in valid_ids)
 
-    def _check_recursion(self, parent=None):
+    def _check_recursion(self, field_name=None):
         """
-        Verifies that there is no loop in a hierarchical structure of records,
-        by following the parent relationship using the **parent** field until a
-        loop is detected or until a top-level record is found.
+        Verifies that there is no loop in a graph of records,
+        by following a given relationship.
+        By default the **parent** field is used as the relationship.
 
-        :param parent: optional parent field name (default: ``self._parent_name``)
+        :param field_name: optional field name (default: ``self._parent_name``)
         :return: **True** if no loop was found, **False** otherwise.
         """
-        if not parent:
-            parent = self._parent_name
+        if not field_name:
+            field_name = self._parent_name
+        field = self._fields.get(field_name)
+        if not field:
+            raise ValueError('invalid field_name: %r' % (field_name,))
+        if not (
+            field.type in ('many2many', 'many2one')
+            and field.comodel_name == self._name
+            and field.store
+        ):
+            raise ValueError('field must be a relation on itself: %r' % (field_name,))
+        if not self.ids:
+            return True
 
-        # must ignore 'active' flag, ir.rules, etc. => direct SQL query
+        # must ignore 'active' flag, ir.rules, etc.
+        # => direct recursive SQL query with cycle detection
+        self.flush_model([field_name])
+        if field.type == 'many2many':
+            relation = field.relation
+            col1 = field.column1
+            col2 = field.column2
+        else:
+            relation = self._table
+            col1 = 'id'
+            col2 = field_name
         cr = self._cr
-        self.flush_model([parent])
-        for id in self.ids:
-            current_id = id
-            seen_ids = {current_id}
-            while current_id:
-                cr.execute(SQL(
-                    "SELECT %s FROM %s WHERE id = %s",
-                    SQL.identifier(parent), SQL.identifier(self._table), current_id,
-                ))
-                result = cr.fetchone()
-                current_id = result[0] if result else None
-                if current_id in seen_ids:
-                    return False
-                seen_ids.add(current_id)
-        return True
+        cr.execute(SQL(
+            """
+            WITH RECURSIVE rel AS (
+                SELECT %(col1)s AS id, %(col2)s AS reach
+                FROM %(rel)s
+                WHERE %(col1)s IN %(ids)s
+                    AND %(col2)s IS NOT NULL
+            UNION
+                SELECT rel.id, t.%(col2)s
```

Comment:
Sorry, but it is wrong 😢. It cannot detect loop like node1 -> node2 -> node3 -> node2, and will cause a dead loop

Reply:
It will not deadloop because UNION will gather all reachable edges and stop.
https://www.db-fiddle.com/f/e7SSMYPasYrA2fXnpr7jgJ/0

It will not detect the loop if you pass only node1 as argument. The function checks the given ids, if you update the relationship for multiple nodes or want to test multiple nodes, you can pass all needed ids.
If it is really required to check loops for reachable but not passed ids, the query below will do the job, but it is slower (I haven't tested exactly by how much).
https://www.db-fiddle.com/f/e7SSMYPasYrA2fXnpr7jgJ/1

Reply:
Oh, yes, you are right! I tried `UNION ALL` long time ago, and forgot to use the `UNION` to avoid deadloop. Since there is not deadloop I think it coud be a good optimization even if the algorithm cannot detect the loop when the provided nodes are not in the loop.
Here is the reason:
In Odoo, these checks should happen after relations changed or new relations created. And all relevant nodes should be added to the check. If these nodes are in a path with loop but not really in the loop, the loop should be created by other old concurrent transactions. We can return True here to allow the business logic for the current transaction to be accepted.

Do you mind I pick these idea and create a new commit/PR, I will add you to the authors

Reply:
I have already updated this PR, so I don't think a new one is necessary. :)